src/excmds.ts

```diff
@@ -2492,52 +2493,28 @@ export async function undo(item = "recent"): Promise<number> {
     const current_win_id: number = (await browser.windows.getCurrent()).id
     const sessions = await browser.sessions.getRecentlyClosed()
 
-    if (item === "tab") {
-        const lastSession = sessions.find(s => {
-            if (s.tab) return true
-        })
-        if (lastSession) {
-            browser.sessions.restore(lastSession.tab.sessionId)
-            return lastSession.tab.id
-        }
-    } else if (item === "window") {
-        const lastSession = sessions.find(s => {
-            if (s.window) return true
-        })
-        if (lastSession) {
-            browser.sessions.restore(lastSession.window.sessionId)
-            return lastSession.window.id
-        }
-    } else if (item === "recent") {
-        // The first session object that's a window or a tab from this window. Or undefined if sessions is empty.
-        const lastSession = sessions.find(s => {
-            if (s.window) {
-                return true
-            } else if (s.tab && s.tab.windowId === current_win_id) {
-                return true
-            } else {
-                return false
-            }
-        })
-
-        if (lastSession) {
-            if (lastSession.tab) {
-                browser.sessions.restore(lastSession.tab.sessionId)
-                return lastSession.tab.id
-            } else if (lastSession.window) {
-                browser.sessions.restore(lastSession.window.sessionId)
-                return lastSession.window.id
-            }
-        }
-    } else if (!isNaN(parseInt(item, 10))) {
-        const sessionId = item
-        const session = sessions.find(s => (s.tab || s.window).sessionId === sessionId)
-        if (session) {
-            browser.sessions.restore(sessionId)
-            return (session.tab || session.window).id
-        }
-    } else {
-        throw new Error(`[undo] Invalid argument: ${item}. Must be one of "tab", "window", "recent"`)
+    // Pick the first session object that is a window or a tab from this window ("recent"), a tab ("tab"), a tab
+    // from this window ("tab_strict"), a window ("window") or pick by sessionId.
+    const predicate =
+        item === "recent"
+            ? s => s.window || (s.tab && s.tab.windowId === current_win_id)
+            : item === "tab"
+            ? s => s.tab
+            : item === "tab_strict"
+            ? s => s.tab && s.tab.windowId === current_win_id
+            : item === "window"
+            ? s => s.window
+            : !isNaN(parseInt(item, 10))
+            ? s => (s.tab || s.window).sessionId === item
+            : s => {
+                  throw new Error("blah")
+              } // this won't throw an error if there isn't anything in the session list, but I don't think that matters
+
+    const session = sessions.find(predicate)
```

Comment:
@fluem do you see what I meant now? : )

Reply:
Great, thanks. Now I understand the syntax.

Reply:
Oh, the ternary `x ? "this" : "that"` isn't what I meant - I meant shifting the argument parsing ("did the user ask for recent, tab_strict, etc.") out of the filter. It would have been totally fine to instead do

```
let predicate
if (item === "recent") predicate = s => ...
else if (item == ...
else throw new Error(...)
```

and, probably, it would have been better to do it that way since it would throw an error even if there weren't any sessions to restore.

Reply:
I like that it does not throw an error, if there are no sessions to restore.